ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/sink/io/Gcs.java

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.ingestion.sink.io;
+
+import com.google.cloud.WriteChannel;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.pubsub.v1.PubsubMessage;
+import com.mozilla.telemetry.ingestion.sink.transform.PubsubMessageToJSONObject;
+import com.mozilla.telemetry.ingestion.sink.transform.PubsubMessageToJSONObject.Format;
+import com.mozilla.telemetry.ingestion.sink.util.BatchWrite;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.UUID;
+
+public class Gcs {
+
+  private Gcs() {
+  }
+
+  public abstract static class Write extends BatchWrite<PubsubMessage, byte[], String, BlobInfo> {
+
+    public static class Ndjson extends Write {
+
+      private final PubsubMessageToJSONObject encoder;
+
+      public Ndjson(Storage storage, long maxBytes, int maxMessages, Duration maxDelay,
+          String batchKeyTemplate, Format format) {
+        super(storage, maxBytes, maxMessages, maxDelay, batchKeyTemplate);
+        this.encoder = new PubsubMessageToJSONObject(format);
+      }
+
+      @Override
+      protected byte[] encodeInput(PubsubMessage input) {
+        return (encoder.apply(input).toString() + "\n").getBytes();
+      }
+    }
+
+    private final Storage storage;
+
+    private Write(Storage storage, long maxBytes, int maxMessages, Duration maxDelay,
+        String batchKeyTemplate) {
+      super(maxBytes, maxMessages, maxDelay, batchKeyTemplate);
+      this.storage = storage;
+    }
+
+    @Override
+    protected String getBatchKey(PubsubMessage input) {
+      return batchKeyTemplate(input);
+    }
+
+    @Override
+    protected Batch getBatch(String batchKey) {
+      return new Batch(storage, batchKey);
+    }
+
+    @VisibleForTesting
+    class Batch extends BatchWrite<PubsubMessage, byte[], String, BlobInfo>.Batch {
+
+      @VisibleForTesting
+      final BlobInfo blobInfo;
+
+      private final WriteChannel writer;
+
+      private Batch(Storage storage, String batchKey) {
+        super();
+        // save blobInfo for the response from this.close
+        final String[] outputPrefixParts = batchKey.split("/", 2);
+        final String bucket;
+        final String keyPrefix;
+        if (outputPrefixParts.length == 2) {
+          bucket = outputPrefixParts[0];
+          keyPrefix = outputPrefixParts[1];
```

Comment:
I will go with adding a trailing `-` if there's no trailing `/`, but doing it in `Sink.main` because  I don't think that `Gcs.Write` should care if the uuid gets appended to letters